Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66021

A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

shunting314 committed:
Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries

A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries"


A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries"


A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries"


A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries"


A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

[ghstack-poisoned]

suo commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -249,10 +249,8 @@ struct __attribute__((visibility("hidden"))) ConcreteInterpreterImpl
-      PyImport_AppendInittab("tensorrt.tensorrt", PyInit_tensorrt);
-    }
+
+    BuiltinRegistry::appendCPythonInittab();
```
Here, it seems we are doing three things:
- Appending to the builtin module list (inittab)
- Sanity check the number of Python modules is what we expect
- Setting `Py_FrozenModules` to our new set of frozen modules.

I wonder if it makes more sense to encapsulate this logic in `BuiltinRegistry`. Overall, the flow looks like:
1. Builtin preinitialization (the 3 steps above)
2. Initialize the python inteprreter
3. Builtin postinitialization (e.g. doing the sys.meta_path trick)

So maybe we should have a `BuiltinRegistry::runPreInitialization` that does step 1 (and checks that `Py_IsInitialized()` is false, etc.) and a `BuiltinRegistry::runPostInitialization`, that does step 3.

That way we can make the interaction between interpreter initialization and `BuiltinRegistry` simple. What do you think?

suo commented:
Cool, overall logic seems fine to me; I mostly left suggestions for ways to clean up the overall flow. But don't take them too literally—if you don't think it would make the code simpler/easier to maintain then we don't have to do them.

shunting314 commented on torch/csrc/deploy/interpreter/interpreter_impl.cpp:
```diff
@@ -288,25 +286,18 @@ struct __attribute__((visibility("hidden"))) ConcreteInterpreterImpl
-        fprintf(
-            stderr, "Error: could not import module 'tensorrt.tensorrt'.\n");
+    {
+      std::string startup(startTemplate);
```
I checked CPython source code, it looks like _Py_PackageContext is used for importing extensions: https://github.com/python/cpython/blob/v3.9.0/Python/importdl.c#L159 . As we import tensorrt.tensorrt as builtin modules, I guess we don't need the _Py_PackageContext trick.

After removing this part, I can still succeed to import tensorrt.tensorrt in the embedding interpreter (using interactive_embedding_interpreter.cpp) and the tests for test_deploy_gpu.cpp pass. So I guess nothing is broken

shunting314 commented on torch/csrc/deploy/interactive_embedded_interpreter.cpp:
```diff
@@ -0,0 +1,16 @@
+/*
```
yep. I have one question. I developed this PR under FBCode. It looks like I can not run setup.py under FBCode since the script requires running under a git repository. Do I have to import this PR to a clean git repository to verify the cmake build? Not sure if ghimport is the tool I need to use

suo replied:
maybe we can follow up with the cmake build; I don't think it's blocking this PR

shunting314 commented on torch/csrc/deploy/interpreter/builtin_registry.h:
```diff
@@ -71,6 +74,9 @@ class BuiltinRegistry {
   static struct _frozen* getAllFrozenModules();
   // call this after all the registration is done.
   static void sanityCheck();
+  static std::vector<std::pair<const char*, void*>> getAllBuiltinModules();
```
This is only used inside this class and in unit test. Let'm change it to a private method and mark unit test as a friend member.

shunting314 committed:
Update on "Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries"


A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries"


A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

[ghstack-poisoned]

suo reviewed (approved):
lgtm :)

shunting314 committed:
Update on "Support Registering a Variable Length List of Builtin Modules for torch::deploy Builtin Libraries"


A builtin library consists of a list of frozen modules and a list of builtin modules. For tensorrt, it's quite simple since we only have a single builtin module tensorrt.tensorrt. But it can be complex for libraries like numpy which contains multiple builtin modules (np.core._multiarray_umath, np.random.mtrand etc.) if we want to add it as a torch::deploy builtin. We enhance the macro that registers builtin libraries to accept a variable length of builtin modules. We can use this macro to register frozentorch, frozenpython, tensorrt for now and can also use it to register libraries like numpy later on.

The enhanced macro now looks as follows. Although we don't need to worry about back-compatibility for now,  but this enhanced version is fully compatible with the previous version. The previous version is just a special case when the library contains no builtin modules.

 ```
REGISTER_TORCH_DEPLOY_BUILTIN(library_name_without_quote, frozen_modules_list,
    builtin_module_name_1, builtin_module_init_function_1, ...,
    builtin_module_name_N, builtin_module_init_function_N)
```

Differential Revision: [D31349390](https://our.internmc.facebook.com/intern/diff/D31349390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31349390/)!

[ghstack-poisoned]
